# Numpy

Numpy is a library for the python programming languange, adding large multi-dimensional arrays and matrices, along with the large collections of the high
level mathematical functions to operate on these arrays
- very fast
- uses vectorization using brodcasting (without using loops)

### 1. Datatypes and Attributes
Numpy has only 1 datatype ndarray (n dimensional array)

In [5]:
import numpy as np

In [6]:
a1 = np.array([1,2,3])            # create a 1d array. It takes python list as parameters 
a1

array([1, 2, 3])

In [7]:
l1 = [4,5,6]
a1 = np.array(l1)

In [8]:
a1

array([4, 5, 6])

In [9]:
a2 = np.array([[1,2,3],            # create 2d array
               [4,5,6]])
a2

array([[1, 2, 3],
       [4, 5, 6]])

In [10]:
a3 = np.array([[[1, 2, 3],
                [4, 5, 6]],        #  create 3d array
               [[7, 8, 9],
                [10, 11, 12]],
               [[13, 14, 15],
                [16, 17, 18]]])

In [12]:
a3

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]])

In [14]:
a1.shape

(3,)

In [15]:
a2.shape

(2, 3)

In [23]:
a3.shape                          # check how many dimensions are there in array

(3, 2, 3)

In [24]:
a1.ndim , a2.ndim, a3.ndim        # check the dimensions of array

(1, 2, 3)

In [25]:
a1.dtype, a2.dtype, a3.dtype       # check the type of data stored in arrray

(dtype('int32'), dtype('int32'), dtype('int32'))

In [26]:
a1.size, a2.size, a3.size         # check the no of elements in the array

(3, 6, 18)

In [27]:
type(a1), type(a2), type(a3)      # check what is the type of the datastructure

(numpy.ndarray, numpy.ndarray, numpy.ndarray)

 ## 2. Creating Numpy Arrays

In [28]:
sample_array = np.array([1,2,3])
print(sample_array)

[1 2 3]


In [30]:
sample_array.dtype

dtype('int32')

In [31]:
ones = np.ones((2,3), dtype=float)    # gives a arrays with given shape and type, all filled with ones  

In [32]:
ones

array([[1., 1., 1.],
       [1., 1., 1.]])

In [33]:
print(ones.dtype)
print(type(ones))

float64
<class 'numpy.ndarray'>


In [34]:
zeros = np.zeros((3,3))   # arrays filled with zeros 

In [35]:
zeros

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [36]:
range_array = np.arange(0, 10, 2)

In [37]:
range_array

array([0, 2, 4, 6, 8])

In [38]:
random_array = np.random.randint(2, 9, size = (3,5))

In [39]:
random_array

array([[8, 3, 3, 3, 6],
       [5, 7, 7, 4, 6],
       [3, 3, 3, 8, 6]])

In [40]:
random_array2 = np.random.random((5,3))

In [41]:
random_array2

array([[0.54545034, 0.05990775, 0.87687031],
       [0.44172388, 0.88969806, 0.0912468 ],
       [0.97318032, 0.93690296, 0.11006074],
       [0.01938437, 0.64097538, 0.20994308],
       [0.74122262, 0.32436457, 0.61371477]])

In [42]:
random_array3 = np.random.rand(3,3)

In [43]:
random_array3

array([[0.33838705, 0.38958052, 0.81603963],
       [0.25403175, 0.5562384 , 0.95392395],
       [0.25304136, 0.03427383, 0.76161778]])

## 3. Viewing arrays and matrices

In [44]:
my_array = np.random.randint(2,9,size=(4,4))
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [45]:
np.unique(my_array)  # returns an array with all elements from the passed array only one time

array([2, 3, 4, 5, 6, 7, 8])

In [46]:
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [48]:
my_array[0], my_array[1], my_array[2], my_array[3]

(array([8, 7, 5, 5]),
 array([4, 4, 8, 2]),
 array([3, 8, 5, 5]),
 array([5, 2, 5, 6]))

In [49]:
my_array[0][0], my_array[0][1], my_array[0][2], my_array[0][3], 

(8, 7, 5, 5)

In [50]:
# We can also use the slicing in numpy arrays. Just like the python
my_array

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [51]:
my_array[0:2]

array([[8, 7, 5, 5],
       [4, 4, 8, 2]])

In [52]:
my_array[2:4]

array([[3, 8, 5, 5],
       [5, 2, 5, 6]])

In [55]:
my_array[0:len(my_array)]

array([[8, 7, 5, 5],
       [4, 4, 8, 2],
       [3, 8, 5, 5],
       [5, 2, 5, 6]])

In [57]:
my_array[0:4:2]

array([[8, 7, 5, 5],
       [3, 8, 5, 5]])

In [58]:
my_array[::-1]

array([[5, 2, 5, 6],
       [3, 8, 5, 5],
       [4, 4, 8, 2],
       [8, 7, 5, 5]])

In [59]:
my_array2 = np.random.randint(10, size = (2,3,4,5))
my_array2

array([[[[3, 0, 1, 9, 8],
         [4, 1, 5, 1, 2],
         [4, 4, 6, 3, 8],
         [9, 8, 1, 5, 9]],

        [[3, 2, 5, 8, 8],
         [3, 7, 0, 7, 3],
         [5, 2, 8, 3, 9],
         [6, 1, 7, 2, 4]],

        [[8, 3, 4, 8, 9],
         [4, 1, 8, 8, 5],
         [8, 2, 9, 0, 0],
         [9, 5, 6, 2, 9]]],


       [[[4, 2, 1, 3, 8],
         [2, 2, 7, 1, 9],
         [6, 3, 4, 1, 4],
         [9, 4, 5, 1, 5]],

        [[1, 2, 1, 1, 2],
         [5, 6, 2, 4, 1],
         [6, 6, 6, 9, 3],
         [1, 1, 3, 4, 2]],

        [[9, 4, 3, 3, 1],
         [3, 4, 9, 9, 0],
         [9, 8, 2, 7, 4],
         [3, 9, 9, 7, 7]]]])

In [60]:
# get the first 4 numbers of the inner most arrays 
my_array2[:, :, :, :1]

array([[[[3],
         [4],
         [4],
         [9]],

        [[3],
         [3],
         [5],
         [6]],

        [[8],
         [4],
         [8],
         [9]]],


       [[[4],
         [2],
         [6],
         [9]],

        [[1],
         [5],
         [6],
         [1]],

        [[9],
         [3],
         [9],
         [3]]]])

 ## 4. Manipulating and Comparing Arrays

### Arithmetic

In [61]:
ones =  np.ones(3)

In [62]:
ones 

array([1., 1., 1.])

In [74]:
a1 = np.array([1,2,3])
a1

array([1, 2, 3])

In [75]:
a1 + ones

array([2., 3., 4.])

In [76]:
a1 - ones

array([0., 1., 2.])

In [77]:
a1 * ones 

array([1., 2., 3.])

In [78]:
a2 = np.array([[1,2,3], [4,5,6]])
a2

array([[1, 2, 3],
       [4, 5, 6]])

In [79]:
a1 * a2

array([[ 1,  4,  9],
       [ 4, 10, 18]])

In [80]:
a3 = np.array([[7,8,9], [11,12,13]])
a3

array([[ 7,  8,  9],
       [11, 12, 13]])

In [81]:
a2 * a3     # mmultiply adjacent elements 

array([[ 7, 16, 27],
       [44, 60, 78]])

In [82]:
a1 * 2

array([2, 4, 6])

In [83]:
a1 / ones 

array([1., 2., 3.])

In [84]:
a2 / a1

array([[1. , 1. , 1. ],
       [4. , 2.5, 2. ]])

In [85]:
a2 // a1 # floor division

array([[1, 1, 1],
       [4, 2, 2]], dtype=int32)

In [86]:
a1 ** 2

array([1, 4, 9], dtype=int32)

In [87]:
a2 ** a1

array([[  1,   4,  27],
       [  4,  25, 216]], dtype=int32)

In [89]:
np.square(a2)

array([[ 1,  4,  9],
       [16, 25, 36]], dtype=int32)

In [95]:
a1

array([1, 2, 3])

In [96]:
ones 

array([1., 1., 1.])

In [97]:
np.add(a1, ones )

array([2., 3., 4.])

In [99]:
np.mod(a2,2) # return remainder after division

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [100]:
a2 % 2

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [103]:
a2 / 2

array([[0.5, 1. , 1.5],
       [2. , 2.5, 3. ]])

In [104]:
a2%2

array([[1, 0, 1],
       [0, 1, 0]], dtype=int32)

In [105]:
np.log(a1)

array([0.        , 0.69314718, 1.09861229])

In [106]:
np.exp(a2)

array([[  2.71828183,   7.3890561 ,  20.08553692],
       [ 54.59815003, 148.4131591 , 403.42879349]])

### Aggregation
Performing the same operation on a number of things 

In [109]:
ls1 = [1,2,3]
type(ls1)

list

In [110]:
sum(ls1)

6

In [111]:
a1 = np.array([1,2,3])
sum(a1)

6

In [112]:
np.sum(a1)

6

- It is advised to perform pythons methods on python datatypes and use numpy's methods on the Numpy arrays

In [117]:
# Create a massive numpy array
massive_array = np.random.random(100000)
massive_array

array([0.90531317, 0.09193317, 0.96397208, ..., 0.47429839, 0.11038701,
       0.289257  ])

In [118]:
massive_array.size

100000

In [119]:
massive_array.dtype

dtype('float64')

In [120]:
type(massive_array)

numpy.ndarray

In [121]:
%timeit sum(massive_array)   #Python's sum method
%timeit np.sum(massive_array) #Numpy sum method

10.4 ms ± 584 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
35.5 µs ± 1.53 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


`%timeit` is the jupyter magic function. It give the time taken by the line to execute.

##### As we can see from the above calculation python `sum()` method take 10.4 milliseconds  while Numpy `np.sum()` method take 35.5 microseconds to perform the sum of 100000 entries.
Pyhton method is 500+ times slower than numpy function. 
<br>
Therefore numpy is great library for ML

In [122]:
np.mean(massive_array)

0.49971779383120823

In [123]:
np.median(massive_array)

0.49847740055854795

In [124]:
np.min(massive_array)   # minimum element

3.932141535711864e-07

In [125]:
np.max(massive_array)   # maximum element

0.9999904286086201

In [126]:
np.std(massive_array)  # standerd deviation

0.2883568024203

In [127]:
np.var(massive_array)   # variance

0.08314964550205992

In [128]:
np.sqrt(massive_array)

array([0.95147947, 0.30320483, 0.9818208 , ..., 0.68869325, 0.33224541,
       0.53782618])

In [129]:
np.square(massive_array)

array([0.81959194, 0.00845171, 0.92924217, ..., 0.22495896, 0.01218529,
       0.08366961])

In [132]:
np.power(massive_array, 3)

array([7.41987383e-01, 7.76992272e-04, 8.95763512e-01, ...,
       1.06697673e-01, 1.34509800e-03, 2.42020220e-02])

## 5. Reshaping and Transposing

In [136]:
arr1 = np.array([[1,2,3],[4,5,6]])
arr1, np.shape(arr1)

(array([[1, 2, 3],
        [4, 5, 6]]),
 (2, 3))

In [137]:
arr2 = np.random.randint(0,10,size = (2,3,3))
arr2, np.shape(arr2)

(array([[[0, 6, 7],
         [0, 4, 8],
         [9, 2, 2]],
 
        [[2, 9, 7],
         [7, 8, 4],
         [5, 5, 3]]]),
 (2, 3, 3))

In [138]:
arr1 * arr2   # error

ValueError: operands could not be broadcast together with shapes (2,3) (2,3,3) 

In [139]:
# Numpy broadcasting rule - two numpy array can only be operated if their shape is same or one of their shape is 1

arr1_reshape = arr1.reshape(2,3,1)
arr1_reshape, np.shape(arr1_reshape)

(array([[[1],
         [2],
         [3]],
 
        [[4],
         [5],
         [6]]]),
 (2, 3, 1))

In [140]:
arr1_reshape * arr2   # now the multiplication is compactible 

array([[[ 0,  6,  7],
        [ 0,  8, 16],
        [27,  6,  6]],

       [[ 8, 36, 28],
        [35, 40, 20],
        [30, 30, 18]]])

In [144]:
arr1.T     # transpose the array

array([[1, 4],
       [2, 5],
       [3, 6]])

In [145]:
arr2.T

array([[[0, 2],
        [0, 7],
        [9, 5]],

       [[6, 9],
        [4, 8],
        [2, 5]],

       [[7, 7],
        [8, 4],
        [2, 3]]])

In [148]:
arr2.shape, arr2.T.shape

((2, 3, 3), (3, 3, 2))

## 6. Comparison Operators

In [150]:
arr1

array([[1, 2, 3],
       [4, 5, 6]])

In [153]:
arr1_reshape

array([[[1],
        [2],
        [3]],

       [[4],
        [5],
        [6]]])

In [158]:
bool_array = arr1_reshape > arr2      # return a boolean array
bool_array

array([[[ True, False, False],
        [ True, False, False],
        [False,  True,  True]],

       [[ True, False, False],
        [False, False,  True],
        [ True,  True,  True]]])

In [159]:
bool_array, bool_array.dtype, type(bool_array)

(array([[[ True, False, False],
         [ True, False, False],
         [False,  True,  True]],
 
        [[ True, False, False],
         [False, False,  True],
         [ True,  True,  True]]]),
 dtype('bool'),
 numpy.ndarray)

In [161]:
arr1_reshape == arr2

array([[[False, False, False],
        [False, False, False],
        [False, False, False]],

       [[False, False, False],
        [False, False, False],
        [False, False, False]]])

In [163]:
arr1_reshape < arr2

array([[[False,  True,  True],
        [False,  True,  True],
        [ True, False, False]],

       [[False,  True,  True],
        [ True,  True, False],
        [False, False, False]]])

## 7. Sorting Arrays

In [171]:
rand_array = np.random.randint(1, 10 ,size = (3,5))

In [172]:
rand_array

array([[3, 8, 7, 6, 3],
       [3, 2, 1, 4, 6],
       [9, 1, 1, 6, 1]])

In [173]:
# Sort the elements in the array in default axis i.e axis = 1 column
np.sort(rand_array)

array([[3, 3, 6, 7, 8],
       [1, 2, 3, 4, 6],
       [1, 1, 1, 6, 9]])

In [175]:
np.sort(rand_array, axis=0) # sort according to axis = 0 row

array([[3, 1, 1, 4, 1],
       [3, 2, 1, 6, 3],
       [9, 8, 7, 6, 6]])

In [179]:
rand_array

array([[3, 8, 7, 6, 3],
       [3, 2, 1, 4, 6],
       [9, 1, 1, 6, 1]])

In [183]:
np.argmax(rand_array, axis=0) # return the indices of maximum element accross the axis

array([2, 0, 0, 0, 1], dtype=int64)

In [184]:
np.argmax(rand_array, axis=1)

array([1, 4, 0], dtype=int64)

## Practical Example - Numpy in Action!!!!

Converting the image into numbers 

<img src='./images/panda.png'/>

In [186]:
# Turn a image into Numpy array
from matplotlib.image import imread

In [187]:
panda = imread('Images/panda.png') # it grab all the pixel and their colour values and store it in a numpy array

In [188]:
print(panda)

[[[0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  ...
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]]

 [[0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  ...
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]]

 [[0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  [0.05490196 0.10588235 0.06666667]
  ...
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]
  [0.16470589 0.12941177 0.09411765]]

 ...

 [[0.13333334 0.07450981 0.05490196]
  [0.12156863 0.0627451  0.04313726]
  [0.10980392 0.05098039 0.03137255]
  ...
  [0.02745098 0.02745098 0.03529412]
  [0.02745098 0.02745098 0.03529412]
  [0.02745098 0.02745098 0.03529412]]

 [[0.13333334 0.07450981 0.05490196]
  [0.12156863 0.0627451  0.04313726]


In [190]:
panda.size, panda.shape, panda.ndim

(24465000, (2330, 3500, 3), 3)

In [191]:
panda[:5]

array([[[0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        ...,
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765]],

       [[0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        ...,
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765]],

       [[0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        ...,
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765],
        [0.16470589, 0.12941177, 0.09411765]],

       [[0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.06666667],
        [0.05490196, 0.10588235, 0.

<img src="Images/car-photo.png"/>

In [193]:
car = imread('Images/car-photo.png')

In [194]:
print(car)

[[[0.5019608  0.50980395 0.4862745  1.        ]
  [0.3372549  0.34509805 0.30588236 1.        ]
  [0.20392157 0.21568628 0.14901961 1.        ]
  ...
  [0.64705884 0.7058824  0.54901963 1.        ]
  [0.59607846 0.63529414 0.45882353 1.        ]
  [0.44705883 0.47058824 0.3372549  1.        ]]

 [[0.44313726 0.43529412 0.40392157 1.        ]
  [0.3137255  0.31764707 0.27450982 1.        ]
  [0.2        0.21176471 0.14117648 1.        ]
  ...
  [0.5058824  0.5372549  0.4117647  1.        ]
  [0.49803922 0.52156866 0.39607844 1.        ]
  [0.4        0.42745098 0.34117648 1.        ]]

 [[0.39607844 0.38039216 0.34117648 1.        ]
  [0.31764707 0.3137255  0.27450982 1.        ]
  [0.28627452 0.29411766 0.24705882 1.        ]
  ...
  [0.44705883 0.45882353 0.32156864 1.        ]
  [0.45882353 0.48235294 0.3529412  1.        ]
  [0.4509804  0.49019608 0.38039216 1.        ]]

 ...

 [[0.47058824 0.57254905 0.6313726  1.        ]
  [0.4392157  0.53333336 0.5882353  1.        ]
  [0.48235

In [195]:
print(type(car))

<class 'numpy.ndarray'>


In [196]:
print(car.size)

991300
